network/registry_test.go

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) 2017 The Alvalor Authors
+//
+// This file is part of Alvalor.
+//
+// Alvalor is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Alvalor is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with Alvalor.  If not, see <http://www.gnu.org/licenses/>.
+
+package network
+
+import (
+    "testing"
+    "github.com/stretchr/testify/assert"
+)
+
+func TestAddSavesPeer(t *testing.T) {
+    reg := registry{ peers : make(map[string]*peer)}
+    addr := "127.0.0.1"
+
+    reg.add(addr, &peer{})
+
+    assert.True(t, reg.has(addr))
+}
+
+func TestRemovesPeer(t *testing.T) {
+    reg := registry{ peers : make(map[string]*peer)}
+    addr := "127.0.0.1"
+    reg.add(addr, &peer{})
+
+    reg.remove(addr)
+
+    assert.False(t, reg.has(addr))
+}
+
+func TestCountsPeers(t *testing.T) {
+    reg := registry{ peers : make(map[string]*peer)}
+
+    reg.add("127.0.0.1", &peer{})
+    reg.add("192.168.4.62", &peer{})
+    reg.add("192.168.77.55", &peer{})
+
+    count := reg.count()
+
+    assert.Equal(t, 3, count)
+}
+
+func TestGetsPeer(t *testing.T) {
+    reg := registry{ peers : make(map[string]*peer)}
+    addr := "127.0.0.1"
+    peerToAdd := &peer{}
+    reg.add(addr, peerToAdd)
+
+    peer := reg.get(addr)
+
+    assert.Equal(t, peerToAdd, peer)
```

Comment:
The assert library will compare the underlying fields of the structs pointed to, and not the actual pointer value.

While I think this should work here, because you would compare nil to a pointer, it's important to note that you are asking for trouble when you don't initialize the peer with some specific fields.

As an example, imagine we would store the peers as values rather than pointers. In this case, the `get` call would return a default initialized peer (with the second return parameter on false) and it would actually equal the peer that you initialized.

In this case, it would make sense to use a peer that has some fields initialized, or to check with `has` first.